### ISSUE TITLE
[iOS][Android][expo-camera] Progress bar when recording video with `maxDuration` option

### ISSUE THREAD

**Issue by graszka22** [expo/expo#23133]
### Minimal reproducible example

https://snack.expo.dev/@blimpapp/recording-progress-example

### Summary

I want to implement a progress bar when user is recording a video with limited duration, so that the user knows how much video they recorder and how much is left. Something like this one on WhatsApp:
![Screenshot_20230627_143328](https://github.com/expo/expo/assets/6280457/5a523ac9-2850-46c4-a0b4-f7288d9b638b)

The problem is the video does not start immediately, there is no callback that the recording has started and there is no way to make this bar synchronised with the video. As a result, when the progress bar ends, there are still a few seconds when the video is still recording and it's confusing to the user.

Expected behaviour:
-  I'm able to start a progress bar when the video starts recording
- The recording ends just in time when progress bar ends

Actual behaviour:
- The recording starts some time after progress bar starts
- The recording still continues for some time after progress bar ends

### Environment

expo-env-info 1.0.5 environment info:
    System:
      OS: macOS 13.3.1
      Shell: 5.9 - /bin/zsh
    Binaries:
      Node: 20.3.0 - /opt/homebrew/bin/node
      Yarn: 1.22.19 - ~/Projects/blimp-reactnative/node_modules/.bin/yarn
      npm: 9.6.7 - /opt/homebrew/bin/npm
    Managers:
      CocoaPods: 1.12.1 - /Users/angelikaserwa/.rbenv/shims/pod
    SDKs:
      iOS SDK:
        Platforms: DriverKit 22.4, iOS 16.4, macOS 13.3, tvOS 16.4, watchOS 9.4
    IDEs:
      Android Studio: 2022.2 AI-222.4459.24.2221.9971841
      Xcode: 14.3.1/14E300c - /usr/bin/xcodebuild
    npmPackages:
      expo: ^48.0.0 => 48.0.4
      react: 18.2.0 => 18.2.0
      react-dom: 18.2.0 => 18.2.0
      react-native: 0.71.11 => 0.71.11
    npmGlobalPackages:
      eas-cli: 3.7.2
      expo-cli: 6.3.0
    Expo Workflow: bare

**Comment by hirbod** [comment]
Ideally we should extend `recordAsync` and pass an `onStart` callback (maybe with an accurate timestamp along with `onEnd`?)

**Comment by baltagih2** [comment]
@hirbod That would be perfect and cover the missing use cases yes!

**Comment by hirbod** [comment]
I think this use case can be implemented easily with `react-native-vision-camera`, cc'ing @mrousavy to verify

**Comment by mrousavy** [comment]
Not yet, I am working on this to make startRecording awaitable.

Update: Done in VisionCamera V5 ✅

**Comment by brentvatne** [comment]
@alanjhughes - is this possible in expo-camera at the moment?

**Comment by alanjhughes** [comment]
@brentvatne no, we don't currently provide this. There wasn't a public api available to accurately do this last time I checked.

**Comment by mrousavy** [comment]
JFYI; There is a public API for this, it works in react-native-vision-camera.

**Comment by alanjhughes** [comment]
@mrousavy I meant iOS does not provide a delegate method. Polling a variable from js is not an iOS provided api

**Comment by mrousavy** [comment]
> Polling a variable from js is not an iOS provided api

It is, the iOS "provided API" is just exposing the currently recorded duration.

If you can then access this synchronously on the UI Thread (which you can in Nitro/VisionCamera), you can implement your progress bar updated in a `useFrameCallback(...)` - which polls this value from natively.

_Something_ has to happen every UI frame for it to update a visual progress bar, and push vs pull makes no difference here for a synchronous getter that is expected to be updated every few milliseconds (aka a time value)

This is how you'd do it in a native app as well.

So if Expo supports accessing the currently recorded duration synchronously and on the UI Thread (in a worklet) like VisionCamera, then this would be a supported feature. I can help give pointers for implementation if you want

**Comment by brentvatne** [comment]
nice, thanks @mrousavy!

**Comment by mrousavy** [comment]
you're welcome ☺️
